### PR TITLE
[SKY-52] Stagger parallel uploads

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

Add a new large file upload option, `staggerPercent`.
  - A percentage from 0-100. When set, parallel uploads are enabled and each chunk is staggered, one after another, instead of all running simultaneously. The stagger percentage is how much of each chunk upload should be finished before the next upload is initiated.

## Dependencies

- https://github.com/SkynetLabs/skynet-js/pull/432
- Changes to `tus-js-client` (PR to come)

### CODE REVIEW

File: src/upload.test.ts [hunk 1]
```diff
@@ -341,10 +351,58 @@ describe("splitSizeIntoChunkAlignedParts", () => {
...
+      [
+        { start: 0, end: 80 * mib },
+        { start: 80 * mib, end: 81 * mib },
+      ],
+    ],
   ];
 
-  it.each(sizesAndChunks)("Should align size '%s' with '%s' parts", (totalSize, partCount, expectedParts) => {
-    const parts = splitSizeIntoChunkAlignedParts(totalSize, partCount);
+  it.each(sizesAndChunks)("('%s', '%s', '%s')", (totalSize, partCount, chunkSize, expectedParts) => {
```
Comment (kwypchlo) [inline src/upload.test.ts:404]: I think the text test name was nice, why did you remove it in favour of just listing the arguments?

**Review by kwypchlo — COMMENTED** [review]
Why are we making this configurable and not even default to some value? I thought that it improves user experience

**Comment by kwypchlo** [comment]
A note, if your pull request depends on other, please change the source branch instead of merging those into each other like this ed479f94b6143f6dab048fe14620e14c5c838213 since that make the review process harder filtering only the necessary changes

File: src/upload.ts [hunk 2]
```diff
@@ -254,32 +279,46 @@ export async function uploadLargeFileRequest(
   // the part-split function. Note that each part has to be chunk-aligned, so we
   // may limit the number of parallel uploads.
   let parallelUploads = 1;
+  let staggerPercent: number | null = null;
```
Comment (peterjan) [inline src/upload.ts:282]: `nit` don't we want to default to something here?

**Review by peterjan — DISMISSED** [review]

File: src/upload.ts (hunk 2 above)
Comment (ChrisSchinnerl) [reply src/upload.ts:282]: 50% would probably be a safe value for most connection speeds.

**Review by ChrisSchinnerl — DISMISSED** [review]

File: src/upload.test.ts (hunk 1 above)
Comment (m-cat) [reply src/upload.test.ts:404]: See discussion on https://github.com/SkynetLabs/skynet-js/pull/432#pullrequestreview-963222675.

**Comment by m-cat** [comment]
> A note, if your pull request depends on other, please change the source branch instead of merging those into each other

@kwypchlo How do I do that? Is that something in `git` or GitHub?

**Review by ro-tex — APPROVED** [review]

**Comment by peterjan** [comment]
> > A note, if your pull request depends on other, please change the source branch instead of merging those into each other
> 
> @kwypchlo How do I do that? Is that something in `git` or GitHub?

if you create a PR you can pick a base branch to diff against and eventually merge it in against, if you edit your PR you'll be able to reselect that and you can easily diff against a branch you're working off of to make it easier for the reviewer

**Review by peterjan — DISMISSED** [review]

**Comment by m-cat** [comment]
[Updated splitSizeIntoChunkAlignedParts to match tus-js-client PR](https://github.com/SkynetLabs/skynet-js/pull/438/commits/b952573e6ff8a6f814c44324a8b6b3eb500adf7d)
[b952573](https://github.com/SkynetLabs/skynet-js/pull/438/commits/b952573e6ff8a6f814c44324a8b6b3eb500adf7d).

Changes made based on feedback in

https://github.com/SkynetLabs/tus-js-client/pull/6

This change should only affect the clarity and readability of the code and not
the functionality.

This function now better matches `buildChunkStaggers` in that PR.

**Review by ro-tex — APPROVED** [review]

File: src/upload.ts [hunk 3]
```diff
@@ -55,6 +60,7 @@ const PORTAL_DIRECTORY_FILE_FIELD_NAME = "files[]";
  * @property [largeFileSize=41943040] - The size at which files are considered "large" and will be uploaded using the tus resumable upload protocol. This is the size of one chunk by default (40 mib). Note that this does not affect the actual size of chunks used by the protocol.
  * @property [errorPages] - Defines a mapping of error codes and subfiles which are to be served in case we are serving the respective error code. All subfiles referred like this must be defined with absolute paths and must exist.
  * @property [numParallelUploads=2] - Used to override the default number of parallel uploads. Disable parallel uploads by setting to 1. Note that each parallel upload must be chunk-aligned so the number of parallel uploads may be limited if some parts would end up empty.
+ * @property [staggerPercent] - A percentage from 0-100. `numParallelUploads` must be more than 1. When set, each chunk is staggered, one after another, instead of all uploads running simultaneously. The stagger percentage is how much of each chunk upload should be finished before the next chunk upload is initiated. Pass `null` to disable staggering.
```
Comment (ro-tex) [inline src/upload.ts:63]: F/U: I don't think docstring is correct. If you pass `null` you'll hit this
```js
throw new Error(`Expected 'staggerPercent' option to be between 0 and 100, was '${opts.staggerPercent}`);
```
It seems that if you want to disable staggering, you need to pass 0.

File: src/upload.ts [hunk 4]
```diff
@@ -281,16 +296,22 @@ export async function uploadLargeFileRequest(
   let splitSizeIntoParts:
     | ((totalSize: number, partCount: number) => Array<{ start: number; end: number }>)
     | undefined = undefined;
+  const staggerPercent = opts.staggerPercent;
```
Comment (ro-tex) [inline src/upload.ts:299]: nit: We don't need this.

File: src/upload.ts [hunk 5]
```diff
@@ -437,25 +465,36 @@ export function splitSizeIntoChunkAlignedParts(
...
-    // the last part in the array if all parts were used.
-    partSizes[Math.min(lastPart + 1, partCount - 1)] += leftover;
+  // The leftover size that must go into the last part.
+  const leftover = totalSize % chunkSize;
+  // If there is non-chunk-aligned leftover, add it.
+  if (leftover > 0) {
+    let lastIndex;
+    if (numFullChunks === 0) {
+      // No parts were visited, so just assign to the last part.
+      lastIndex = partCount - 1;
```
Comment (ro-tex) [inline src/upload.ts:491]: Q: If no parts were visited, that means that the total size of the upload is under the chunk size. That would mean that all parts, except the last one, will be empty. That's perfectly fine by me, as long as TUS won't complain about it. Is this the case, both in terms of part sizes and TUS behaviour?

**Review by ro-tex — APPROVED** [review]

**Review by peterjan — APPROVED** [review]

File: src/upload.ts (hunk 3 above)
Comment (m-cat) [reply src/upload.ts:63]: Don't we guard against that in the 3rd line here?

```ts
if (
  opts.staggerPercent !== undefined &&
  opts.staggerPercent !== null &&
  (opts.staggerPercent < 0 || opts.staggerPercent > 100)
) {
  throw new Error(`Expected 'staggerPercent' option to be between 0 and 100, was '${opts.staggerPercent}`);
}
```

File: src/upload.ts (hunk 4 above)
Comment (m-cat) [reply src/upload.ts:299]: Ended up doing a slight refactor here because if we end up using `parallelUploads: 1` then this has to be set to null.

File: src/upload.ts (hunk 5 above)
Comment (m-cat) [reply src/upload.ts:491]: Honestly I've never run into this because we disable parallel uploads when we're under a chunk. I don't know how it behaves, but this should just throw as it's an unexpected flow path.

Comment (m-cat) [reply src/upload.ts:491]: PS having e.g. two parts and only using one of them is fine, this has been tested. I'll document that in a comment.

File: src/upload.ts (hunk 3 above)
Comment (ro-tex) [reply src/upload.ts:63]: Maybe I misunderstood the code. Is throwing this error our way of disabling staggering? It seemed like it would fail the upload with a 400. When I read "Pass `null` to disable staggering." I imagined that the upload will succeed and the chunks will go all together, i.e. 0% staggering. 

There also seem to be no tests for this and it would be nice to cover it.

**Review by peterjan — APPROVED** [review]

File: src/upload.ts (hunk 3 above)
Comment (m-cat) [reply src/upload.ts:63]: The error is not thrown if `staggerPercent` is `null`. If it is `null` then we just don't stagger, i.e. do a regular parallel upload. I have just tested this to verify this behavior.

> There also seem to be no tests for this and it would be nice to cover it.

Unfortunately, setting up integration tests for tus is a nontrivial task. Looks like there is already an issue to track it; I added some detail:

https://linear.app/skynetlabs/issue/SKY-304/add-tus-testing-to-integration-tests-health-checks

Unit tests are also nontrivial, because there are many requests that would have to be precisely mocked to emulate the protocol.

Comment (ro-tex) [reply src/upload.ts:63]: Oh, boy. I just fail at reading again. Of course it doesn't throw when it's a `null` - the comparison is `!==` and not `===`.

**Review by ro-tex — APPROVED** [review]
